pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -70,61 +71,92 @@ func (c *Controller) ConfigDescriptor() model.ConfigDescriptor {
 }
 
 // List returns all the config that is stored by type and namespace
-func (c *Controller) List(typ, namespace string) ([]model.Config, error) {
+// if namespace is empty string it returns config for all the namespaces
+func (c *Controller) List(typ, namespace string) (out []model.Config, err error) {
 	_, ok := c.ConfigDescriptor().GetByType(typ)
 	if !ok {
 		return nil, fmt.Errorf("list unknown type %s", typ)
 	}
 	data, exists := c.configStore[typ]
-	if !exists {
-		log.Infof("list config not found for the type %s\n", typ)
-		return nil, nil
+	if exists && len(data) == 0 {
+		return out, nil
 	}
-	out := make([]model.Config, 0, len(c.configStore[typ]))
+
 	if namespace == "" {
-		for _, ns := range data {
-			ns.Range(func(key, value interface{}) bool {
-				out = append(out, value.(model.Config))
-				return true
-			})
+		c.configStoreMu.Lock()
+		byType, ok := c.configStore[typ]
+		c.configStoreMu.Unlock()
+		if ok {
+			for _, byNamespace := range byType {
+				for _, config := range byNamespace {
+					out = append(out, config)
+				}
+			}
+			return out, nil
 		}
-	} else {
-		ns, exists := data[namespace]
-		if !exists {
-			return nil, nil
+	}
+
+	if byNamespace, ok := data[namespace]; ok {
+		for _, config := range byNamespace {
+			out = append(out, config)
 		}
-		ns.Range(func(key, value interface{}) bool {
-			out = append(out, value.(model.Config))
-			return true
-		})
 	}
 	return out, nil
 }
 
 // Apply receives changes from MCP server and creates the
 // corresponding config
 func (c *Controller) Apply(change *mcpclient.Change) error {
+	typ := extractType(change.MessageName)
+	// is this a realistic case to test?
```

Comment:
An empty change set is realistic. It would indicate all of resources of a specific type were removed. That said, you shouldn't return early. We need to clear the `innerStore` for said type so the change is reflected to pilot.

Reply:
That makes sense, going to test drive that logic.